theano/compile/builders.py

```diff
@@ -115,40 +210,203 @@ def __hash__(self):
         # TODO: use internal variables in hash
         return hash(type(self))
 
-    def make_node(self, *inputs):
-        for input, type in zip(inputs, self.input_types):
-            if not type == input.type:
-                raise TypeError("Wrong type, expected %s but got %s" %
-                                (type, input.type))
-        return gof.Apply(self,
-                         list(inputs) + self.shared_inputs,
-                         [type() for type in self.output_types])
+    def __str__(self):
+        name = self.__class__.__name__ if self.name is None else self.name
+        is_inline = self.is_inline
+        return '%(name)s{inline=%(is_inline)s}' % locals()
+
+    def _recompute_grad_op(self):
+        if isinstance(self._grad_op, OpFromGraph):
+            self._grad_op_is_cached = True
+            return
+        output_grads = [out_t() for out_t in self.output_types]
+        if self._grad_op is None:
+            self._grad_op = []
+
+        # we need to convert a list/function into an OfG instance
+        if isinstance(self._grad_op, list):
+            goverrides_l = self._grad_op
+            if len(goverrides_l) > len(self.local_inputs):
+                raise ValueError(
+                    'Can override %d gradients at most, got %d' % (
+                        len(self.local_inputs), len(goverrides_l)),
+                    self.goverrides_l)
+            if len(goverrides_l) < len(self.local_inputs):
+                goverrides_l += [None] * (
+                    len(self.local_inputs) - len(goverrides_l))
+            wrt_l = [lin for lin, gov in
+                     izip(self.local_inputs, goverrides_l) if not gov]
+            # compute non-overriding downsteam grads from upstreams grads
+            # it's normal some input may be disconnected, thus the 'ignore'
+            gdefaults = iter(theano.gradient.grad(
+                cost=None,
+                known_grads=dict(izip(self.local_outputs, output_grads)),
+                wrt=wrt_l,
+                disconnected_inputs='ignore') if wrt_l else [])
```

Comment:
Oh, OK, I thought `wrt_l` was iterated through.
It may be worth setting `return_disconnected='None'` and `null_gradients='return'` however.